Kaituo forecasting (#10133)
opensearch-project/documentation-website: _observing-your-data/forecast/getting-started.md

## Step 1: Define a forecaster

A **forecaster** represents a single forecasting task. You can create multiple forecasters to run in parallel, each analyzing a different data source. Follow these steps to define a new forecaster:

1. In the **Forecaster list** view, choose **Create forecaster**.

2. Define the data source by entering the following information:
   * **Name** – Provide a unique, descriptive name, such as `requests-10min`.  
   * **Description** – Summarize the forecaster's purpose, for example: `Forecast total request count every 10 minutes`.  
   * **Indexes** – Select one or more indexes, index patterns, or aliases. Remote indexes are supported through cross-cluster search (`cluster-name:index-pattern`). For details, see [Cross-cluster search]({{site.url}}{{site.baseurl}}/search-plugins/cross-cluster-search/). If the Security plugin is enabled, refer to [Selecting remote indexes with fine-grained access control]({{site.url}}{{site.baseurl}}/observing-your-data/forecast/security/#selecting-remote-indexes-with-fine-grained-access-control) in the forecasting security documentation.

> 3. (Optional) Choose **Add data filter** to set a **Field**, **Operator**, and **Value**, or choose **Use query DSL** to define a [boolean query]({{site.url}}{{site.baseurl}}/query-dsl/compound/bool/). The following example uses a DSL filter to match three URL paths:

Author: ```suggestion
3. (Optional) Choose **Add data filter** to set a **Field**, **Operator**, and **Value**, or choose **Use query DSL** to define a [Boolean query]({{site.url}}{{site.baseurl}}/query-dsl/compound/bool/). The following example uses a DSL filter to match three URL paths:
```

   * **Indexes** – Select one or more indexes, index patterns, or aliases. Remote indexes are supported through cross-cluster search (`cluster-name:index-pattern`). For details, see [Cross-cluster search]({{site.url}}{{site.baseurl}}/search-plugins/cross-cluster-search/). If the Security plugin is enabled, refer to [Selecting remote indexes with fine-grained access control]({{site.url}}{{site.baseurl}}/observing-your-data/forecast/security/#selecting-remote-indexes-with-fine-grained-access-control) in the forecasting security documentation.

3. (Optional) Choose **Add data filter** to set a **Field**, **Operator**, and **Value**, or choose **Use query DSL** to define a [Boolean query]({{site.url}}{{site.baseurl}}/query-dsl/compound/bool/). The following example uses a DSL filter to match three URL paths:

     ```json
     {
       "bool": {
         "should": [
           { "term": { "urlPath.keyword": "/domain/{id}/short" } },
           { "term": { "urlPath.keyword": "/sub_dir/{id}/short" } },
           { "term": { "urlPath.keyword": "/abcd/123/{id}/xyz" } }
         ]
       }
     }
    ```

> 4. Under **Timestamp field**, select the field that stores event time.

Author: ```suggestion
4. Under **Timestamp field**, select the field that stores the timestamps.
```

| `plugins.forecast.forecast_result_history_retention_period` | Duration to retain forecast results. Supports duration formats such as `7d`, `90d`. | `30d` |

### Specifying a custom results index

You can store forecast results in a custom index by selecting **Custom index** and providing an alias name, such as `abc`. The plugin creates an alias like `opensearch-forecast-result-abc` that points to the backing index (for example, `opensearch-forecast-result-abc-history-2024.06.12-000002`).

To manage permissions, use hyphenated namespaces. For example, assign `opensearch-forecast-result-financial-us-*` to roles for the `financial` department’s `us` group.
{: .note } If the Security plugin is enabled, ensure appropriate [permissions are configured]({{site.url}}{{site.baseurl}}/observing-your-data/forecast/security/#custom-result-index-permissions).

### Flattening nested fields

If your custom results index includes nested fields, enable the **Flattened custom result index** to simplify aggregation and visualization.

> 

Author: Documents

This creates a separate index prefixed with the custom index and forecaster name (for example, `opensearch-forecast-result-abc-flattened-test`) and attaches an ingest pipeline using a [Painless script](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/resources/scripts/flatten-custom-result-index-painless.txt) to flatten nested data.

If you later disable this option, the associated ingest pipeline is removed.

Use [Index State Management]({{site.url}}{{site.baseurl}}/im-plugin/ism/index/) to manage rollover and deletion of flattened result indexes.

### Custom result index lifecycle management

The plugin triggers a rollover for custom result indexes when any of the following conditions are met:

| Parameter | Description | Type | Unit | Default | Required |
|----------|-------------|------|------|---------|----------|
| `result_index_min_size` | The minimum total primary shard size before rollover. | Integer | MB | `51200` (50 GB) | No |
| `result_index_min_age` | The minimum index age to trigger rollover. | Integer | Days | `7` | No |

> | `result_index_ttl` | The minimum age before deletion. | Integer | Days | `60` | No |

Author: ```suggestion
| `result_index_ttl` | The minimum amount of time before rolled-over indexes are deleted | Integer | Days | `60` | No |
```

| **Distance to threshold value** | Filters forecasts by a numeric threshold (>, <, ≥, ≤) and then orders the remainder by how far they sit from that threshold. | Investigate entities that breach—or nearly breach—an SLA or business KPI, such as "show anything forecast to exceed 10,000 requests". |

If the forecaster monitors five or fewer entities, **Multi-series view** displays all of them. When there are more than five, the view reranks them dynamically each time you change the sort method or adjust the threshold, ensuring that the most relevant series stay in focus.

To focus on a specific subset of entities, switch **Filter by** to **Custom query** and enter a query DSL query. The following example shows entities where the `host` equals `server_1`:

```json
{
  "nested": {
    "path": "entity",
    "query": {
      "bool": {
        "must": [
          { "term":     { "entity.name":  "host"     } },
          { "wildcard": { "entity.value": "server_1" } }
        ]
      }
    }
  }
}
```

> Next, select a sort method, such as **maximum forecast value in descending** order, and select **Update visualization**. The chart updates to show only the forecast series for `host:server_1`, ranked according to your selected criteria.

Author: ```suggestion
Next, select a sort method, such as **Maximum value within the horizon**, and select **Update visualization**. The chart updates to show only the forecast series for `host:server_1`, ranked according to your selected criteria.
```
